classes/backup/class-backup-status.php

```diff
@@ -2,44 +2,209 @@
+		}
+
+		// If the backup is started and we don't support file locks then we have to assume we're still running.
+		return true;
+	}
+
+	/**
+	 * If the running file exists but isn't locked then the thread that
+	 * the backup process is running in must have been killed.
+	 *
+	 * You should only be running this command from a separate thread.
+	 *
+	 * @return boolean Whether the backup process has crashed or not.
+	 */
+	public function has_crashed() {
+		return ( $this->is_started() && ! $this->is_running() );
+	}
+
+	/**
+	 * Handle a process that's previously crashed.
+	 *
+	 * Delete the partially created backup if it exists and then
+	 * run the standard cleanup tasks and set an error message for the user.
+	 *
+	 * @return bool Whether the crash was handled or not.
+	 */
+	public function cleanup_after_crash() {
+
+		if ( ! $this->has_crashed() ) {
+			return false;
+		}
+
+		if ( file_exists( trailingslashit( Path::get_path() ) . $this->get_backup_filename() ) ) {
+			unlink( trailingslashit( Path::get_path() ) . $this->get_backup_filename() );
+		}
+
+		$this->finish();
+
+		// If we already have a fatal error then let's not stomp on it.
+		if ( Notices::get_instance()->get_notices( 'backup_failed' ) ) {
+			$message = __( 'Your last backup failed. The backup process was killed before it could complete. Please contact your host for assistance or try excluding more files.', 'backupwordpress' );
+			Notices::get_instance()->set_notices( 'backup_failed', array( $message ), true );
+		}
+
+		return true;
+	}
+
+	/**
+	 * Catch fatal errors and react accordingly.
+	 *
+	 * Hooked into the shutdown action. If we've shutdown because of a Fatal error
+	 * then we cleanup and set an error message for the user.
+	 */
+	public function catch_fatals() {
+
+		$error = error_get_last();
+
+		if ( empty( $error ) ) {
+			return;
+		}
+
+		if ( ! isset( $error['type'] ) || ! defined( 'E_ERROR' ) || E_ERROR !== $error['type'] ) {
+			return;
+		}
+
+		if ( file_exists( trailingslashit( Path::get_path() ) . $this->get_backup_filename() ) ) {
+			unlink( trailingslashit( Path::get_path() ) . $this->get_backup_filename() );
+		}
+
+		$this->finish();
+
+		$message = sprintf(
+			__( 'Your last backup failed. The backup process encountered an error before it could complete. The error was %s. Please contact your host for assistance or try excluding more files.', 'backupwordpress' ),
+			'<code>' . esc_html( $error['message'] ) . '</code>'
+		);
+		Notices::get_instance()->set_notices( 'backup_failed', array( $message ), true );
+	}
+
+	/**
+	 * Get the filepath for the backup file we're tracking.
+	 *
+	 * @return string The path to the backup file.
+	 */
+	public function get_backup_filename() {
+
+		if ( $this->is_started() ) {
+			$status = json_decode( file_get_contents( $this->get_status_filepath() ) );
+
+			if ( ! empty( $status->filename ) ) {
+				$this->filename = $status->filename;
+			}
+		}
+
+		return $this->filename;
```

Comment:
Seems a bit weird to always return this even if we haven't actually gotten a filename from the file